### PR TITLE
LPS-80771 Fix bugs of changing types

### PR DESCRIPTION
**Pull request by alinedoleron**
Related Issue: https://issues.liferay.com/browse/LPS-80771

### CODE REVIEW

**Comment by liferay-forms** [comment]
Hi, you don’t need to run the CI commands anymore, I’ll do it for you and let you know if any issue happens.

**Comment by liferay-forms** [comment]
ci:test:sf

**Comment by liferay-forms** [comment]
ci:test:relevant

**Comment by alinedoleron** [comment]
ci:test:relevant

**Comment by liferay-forms** [comment]
Hey @alinedoleron, can you please check if these failures are related to your changes? Thanks

**Comment by natocesarrego** [comment]
@carolmariaabb, once Alininha is on vacation, can you please check if the reported test failure is related to this pull?

Thanks in advance.

**Comment by carolmariaabb** [comment]
> @carolmariaabb, once Alininha is on vacation, can you please check if the reported test failure is related to this pull?
> 
> Thanks in advance.

Sure, @natocesarrego !

**Comment by carolmariaabb** [comment]
Just started reviewing :)<br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by carolmariaabb** [comment]
Hi @natocesarrego, I tested it locally on this branch and on Master and in both cases there is a problem that I couldn't figure out how to solve it properly. 

For some reason, the test is taking too long at [FormFields.macro#L586](https://github.com/liferay/liferay-portal/blob/6e486ffaee3faa9e0c7495781ebeab5518d6a94b/portal-web/test/functional/com/liferay/portalweb/macros/FormFields.macro#L586).

Because of that, at some point, the session is expired and the user is unable to edit/save the Form, as you can see in this image from the Poshi Report:

image

Therefore, the failure is not related to this PR but we need to do an investigation.

**Comment by natocesarrego** [comment]
Thanks for your clarification, @carolmariaabb. I've just checked the [history](https://testray.liferay.com/home/-/testray/case_results/1331590339/history?tab=result&hideFilterPin=true) of this test and indeed this same test failed with the same cause as this pull on Oct, 26. I think we should warn @cvanut and @lucasneves91.

File: portal-web/test/functional/com/liferay/portalweb/tests/enduser/webformsanddatalists/forms/FormRules.testcase [hunk 1]
```diff
@@ -35,6 +35,75 @@ definition {
...
+		FormRules.configureActionOptions(
+			action = "Require",
+			actionRowNumber = "1",
+			actionTarget = "Text Field 2");
+
+		FormsAdminNavigator.gotoFormBuilder();
+
+		AssertVisible(locator1 = "Form#ADD_FIELD");
+
+		FormsAdminNavigator.gotoRules();
```
Comment (natocesarrego) [inline portal-web/test/functional/com/liferay/portalweb/tests/enduser/webformsanddatalists/forms/FormRules.testcase:86]: @alinedoleron, is this really necessary?

File: portal-web/test/functional/com/liferay/portalweb/tests/enduser/webformsanddatalists/forms/FormRules.testcase [hunk 2]
```diff
@@ -35,6 +35,75 @@ definition {
...
+
+		FormRules.save();
+
+		FormsAdminNavigator.gotoFormBuilder();
+
+		Form.save();
+
+		FormsAdminNavigator.gotoEditField(fieldName = "Text Field");
```
Comment (natocesarrego) [inline portal-web/test/functional/com/liferay/portalweb/tests/enduser/webformsanddatalists/forms/FormRules.testcase:94]: @alinedoleron, isn't `Text Field` the field label?

File: portal-web/test/functional/com/liferay/portalweb/tests/enduser/webformsanddatalists/forms/FormRules.testcase [hunk 3]
```diff
@@ -35,6 +35,75 @@ definition {
...
+
+		Form.changeFieldTypeModalConfirmation(selectOption = "Select from List");
+
+		FormFields.viewTextLabel(
+			fieldLabel = "Text Field",
+			fieldName = "Text Field");
```
Comment (natocesarrego) [inline portal-web/test/functional/com/liferay/portalweb/tests/enduser/webformsanddatalists/forms/FormRules.testcase:100]: @alinedoleron, isn't `Text Field` the field label?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js [hunk 4]
```diff
@@ -86,7 +86,7 @@ const getFieldType = (fieldName, pages) => {
 
 const optionBelongsToRule = (condition, options) => {
 	return options.some(
-		(option) => option.value === condition.operands[1].value
+		(option) => option.value === condition.operands[1]?.value
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js:89]: Great job, @alinedoleron! 👏 I think you got back to your JS studies, right? 🙂

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js [hunk 5]
```diff
@@ -310,16 +328,16 @@ const fieldNameBelongsToCondition = (fieldName, conditions) => {
 		.some((fieldFound) => fieldFound === true);
 };
 
-const findRuleByFieldName = (fieldName, rules) => {
+const findRuleByFieldName = (fieldName, rules, pages) => {
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js:331]: @alinedoleron, please list the parameters in alphabetic order.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js [hunk 6]
```diff
@@ -310,16 +328,16 @@ const fieldNameBelongsToCondition = (fieldName, conditions) => {
...
 		(rule) =>
-			fieldNameBelongsToAction(fieldName, rule.actions) ||
+			fieldNameBelongsToAction(fieldName, rule.actions, pages) ||
 			fieldNameBelongsToCondition(fieldName, rule.conditions)
 	);
 };
 
-const findInvalidRule = (rule) => {
-	return findRuleByFieldName('', [rule]);
+const findInvalidRule = (rule, pages) => {
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-builder/src/main/resources/META-INF/resources/js/components/RuleBuilder/RulesSupport.es.js:339]: @alinedoleron, please list the parameters in alphabetic order.

**Review by natocesarrego — CHANGES_REQUESTED** [review]

**Comment by alinedoleron** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Pull request closed without merging**